Add code samples to hooks-and-observers release notes (#1437)
bevyengine/bevy-website: release-content/0.14/release-notes/10756_hooks_and_observers.md

If we use a pull-based approach (`RemovedComponents` is the most natural here), there can be gaps (within a single frame) between the entity being despawned and the `Target` component being updated.
This can lead to all sorts of bizarre bugs: fun for speedrunners, but not for game developers.

Let's see what this looks like with a hook on `Targetable`:

```rust
// Rather than a derive, let's configure the hooks with a custom implementation of Component
impl Component for Targetable {
    const STORAGE_TYPE: StorageType = StorageType::Table;

    fn register_component_hooks(hooks: &mut ComponentHooks) {

>         // Whenever this component is removed, or an entity with this component is respawned...

Reviewer: ```suggestion
        // Whenever this component is removed, or an entity with this component is despawned...
```

        hooks.on_remove(|mut world, targeted_entity, _component_id|{
            // Grab the data that's about to be removed
            let targetable = world.get::<Targetable>(targeted_entity).unwrap();
            for targeting_entity in targetable.targeted_by {
                // Track down the entity that's targeting us
                let mut targeting = world.get::<Targeting>(targeting_entity).unwrap();
                // And clear its target, cleaning up any dangling references
                targeting.0 = None;
            }
        })
    }
}
```

Because adding and removing components can only be done in the context of exclusive world access, hooks are always run *immediately*, leaving no opportunity for desynchronization.

> By contrast, observers are a flexible tool intended for gameplay logic.

Reviewer: ```suggestion
By contrast, observers are a flexible tool intended for higher level application logic.
```

They can listen to the same lifecycle events as hooks, but can also respond to custom, user-defined triggers.
Their advantages over buffered events are clearest when you're targeting a specific entity,
when combined with commands that emit triggers (to avoid ever entering a bad state),
or when you're taking advantage of observers' ability to emit triggers which are then immediately processed, chaining recursively.

Let's examine the API through a simple gameplay-flavored example:

```rust
// Any event type can be used as a trigger
#[derive(Event)]
struct DealDamage {
    damage: u8
}

#[derive(Event)]
struct LoseLife {
    life_lost: u8,
}

#[derive(EVent)]
struct PlayerDeath;

// Observers are stored as components on entities,
// and can be set up to watch specific entities.

> fn spawn_player(mut commands: Commands){

Reviewer 2: The other functions are missing this space as well.
```suggestion
fn spawn_player(mut commands: Commands) {
```

    let defense = query.get(trigger.entity).unwrap_or_default();
    let damage = trigger.event().damage;
    let life_lost = damage.0.saturating_sub(defense.0);
    // Observers can be chained into each other, by sending more triggers using commands
    commands.trigger_targets(trigger.entity)
}

fn respond_to_losing_life(trigger: Trigger<LoseLife>, mut life_query: Query<&mut Life>, player_query: Query<Entity, With<Player>>, mut commands: Commands) {
    let mut life = life_query.single_mut(trigger.entity);
    let life_lost = trigger.event().life_lost;
    life.0 = life.0.saturating_sub(life_lost);

    if player_query.contains(trigger.entity){
        // Triggers can be sent globally, targeting no entity in particular
        commands.trigger(PlayerDeath);
    }
}

app
    .add_systems(Startup, spawn_player)
    .add_systems(Update, attack_player)
    // Similarly, observers can also be registered globally, listening to any matching event,
    // regardless of its entity target
    .observe(|_trigger: Trigger<PlayerDeath>, app_exit: EventWriter<AppExit>| {
        println!("You died. Game over!")
        app_exit.send_default();    

> });

Reviewer: ```suggestion
    });
```
